### DISCUSSION TITLE
`IOException` *Invalid function* unless I wait a bit before accessing a file

### DISCUSSION THREAD

**Discussion by lostmsu** [dokan-dev/dokan-dotnet#356] [General]
I am guessing same as `ERROR_INVALID_FUNCTION` code 1 Windows System Error. Shows as "INVALID DEVICE REQUEST" in procmon.

This does not happen on Mirror (at least not with trivial repro steps), but does happen on my overlay FS that combines another read-only FS with a regular directory where any writes go.

I try to read a file immediately after mounting the FS (`DokanInstanceBuilder.Build` call). `CreateFile` succeeds, but `ReadFile` fails with this error according to procmon and the `IOException`. Roughly, the code is:

```csharp
var dokanInstance = builder.Build(fs);
await using var stream = new FileInfo(path).OpenRead();
await JsonSerializer.DeserializeAsync<SomeClass>(stream, cancel)
                    .ConfigureAwait(false)!;
                    // IOException: Incorrect function
```

The funky thing is: during this time neither `CreateFile` nor `ReadFile` breakpoints in my FS are hit.

Now if I add a 15 seconds delay after mounting, the call succeeds, and my breakpoints in `CreateFile` and `ReadFile` also get hit.

```csharp
var dokanInstance = builder.Build(fs);
await Task.Delay(TimeSpan.FromSeconds(15)).ConfigureAwait(false);
await using var stream = new FileInfo(path).OpenRead();
await JsonSerializer.DeserializeAsync<SomeClass>(stream)
                    .ConfigureAwait(false)!;
                    // OK
```
Procmon trace: [IncorrectFunction.zip](https://github.com/dokan-dev/dokan-dotnet/files/15029470/IncorrectFunction.zip)

![InvalidFunction](https://github.com/dokan-dev/dokan-dotnet/assets/239520/27d654c3-1c43-4b1f-8b46-3a9a9fa4e6ba)

**Comment by lostmsu** [comment]
UPD.: While failures occur, there are no logs from `CreateFileProxy`. If I continue retrying, the first one shows up and then everything since that point succeeds.

**Comment by Olof-Lagerkvist** [comment]
You need to wait until the drive is ready. I usually do that in a loop similar to this: 

```
var driveInfo = new DriveInfo(mountPoint);
while (!driveInfo.IsReady)
{
    await Task.Delay(500);
}
```

**Reply by lostmsu** [reply]
Does this work for mount points that are directories (as opposed to drives)? AFAIR `DriveInfo` will only use the drive portion of `mountPoint`.

UPD. I checked, it does not. `IsReady` just does `Directory.Exists`.

**Reply by Olof-Lagerkvist** [reply]
That's correct unfortunately. You would need to find another way for directory mount points. For example, try to enumerate files in the mount points directory. That will throw an exception until it is ready.

Or, try another approach and let your file system implementation notify when it gets first calls and know that way that it is ready.

**Reply by lostmsu** [reply]
I created https://github.com/dokan-dev/dokan-dotnet/issues/357 because I think this behavior is surprising.
